File: SDK Examples/Examples/C#/Counter_BufferedPWMOutput/BufferedPWMOutputForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Threading;
using Automation.BDaq;

namespace Counter_BufferedPWMOutput
{
   public partial class BufferedPWMOutputForm : Form
   {
      public BufferedPWMOutputForm()
      {
         InitializeComponent();
      }

      private void HandleError(ErrorCode err)
      {
         if (err != ErrorCode.Success)
         {
            MessageBox.Show("Sorry ! Some errors happened, the error code is: " + err.ToString(), "BufferedPWMOutput");
         }
      }

      private void BufferedPwModulatorForm_Load(object sender, EventArgs e)
      {
         //The default device of project is demo device, users can choose other devices according to their needs. 
         // no device is selected
         if (!bufferedPwModulatorCtrl1.Initialized)
         {
            MessageBox.Show("No device be selected or device open failed!");
            this.Close();
            return;
         }

         //set title of the form.
         string description = bufferedPwModulatorCtrl1.SelectedDevice.Description;
         this.Text = "Buffered PWM Output(" + description + ")";
         Stop.Enabled = false;
      }

      private static void ShowErrorMessage(Exception e)
      {
         string errorInfo;
         errorInfo = "There's some error happened, the error information: ";
         MessageBox.Show(errorInfo + e.Message);
      }

      private void Start_Click(object sender, EventArgs e)
      {
         ErrorCode ret = ErrorCode.Success;
         // Enable execution status
         ExecutionStatus.Enabled = true;
         int channelCountMax = bufferedPwModulatorCtrl1.Features.ChannelCountMax;
         PoChannel[] poChannel = bufferedPwModulatorCtrl1.Channels;
         int sampleCount = bufferedPwModulatorCtrl1.Samples;
         int intervelCount = bufferedPwModulatorCtrl1.IntervalCount;
         PulseWidth[] dataBuf;
         dataBuf = new PulseWidth[sampleCount];
         byte[] channelEnable;
         channelEnable = new byte[channelCountMax];
         channelEnable = bufferedPwModulatorCtrl1.ChannelMap;

         for (int i = 0; i < sampleCount; i++)
         {
            if ((i & 1) == 0)
            {
               dataBuf[i].HiPeriod = Double.Parse(data1_HiPeriod.Text);
               dataBuf[i].LoPeriod = Double.Parse(data1_LoPeriod.Text);
            }
            else
            {
               dataBuf[i].HiPeriod = Double.Parse(data2_HiPeriod.Text);
               dataBuf[i].LoPeriod = Double.Parse(data2_LoPeriod.Text);
            }
         }

         bufferedPwModulatorCtrl1.IntervalCount = intervelCount;
         bufferedPwModulatorCtrl1.Samples = sampleCount;

         for (int i = 0; i < channelCountMax; i++ )
         {
            if(channelEnable[i] == 1)
            {
               bufferedPwModulatorCtrl1.SetData(i, sampleCount, dataBuf);
            }
         }

         ret = bufferedPwModulatorCtrl1.Prepare();
         if (ret != ErrorCode.Success)
         {
            HandleError(ret);
            return;
         }
        
         try
         {
            bufferedPwModulatorCtrl1.Enabled = true;
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }

         Stop.Enabled = true;
         Start.Enabled = false;
      }

      private void Stop_Click(object sender, EventArgs e)
      {
         try
         {
            bufferedPwModulatorCtrl1.Enabled = false;
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }
         // Disable execution status
         ExecutionStatus.Enabled = false;
         Start.Enabled = true;
         Stop.Enabled = false;
      }
   }
}

File: SDK Examples/Examples/VB.Net/Counter_BufferedPWMOutput/BufferedPWMOutputForm.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Threading
Imports Automation.BDaq


Partial Public Class BufferedPWMOutputForm
   Inherits Form
   Private m_sampleCount As UInteger
   Private m_intervelCount As UInteger
   Private m_channelEnable As Byte()
   Private m_channelCountMax As Integer

   Public Sub New()
      InitializeComponent()
   End Sub

   Private Sub BufferedePWMOutputForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
      ' no device is selected
      If Not BufferedPwModulatorCtrl1.Initialized Then
         MessageBox.Show("Please select a device in control property!")
         Me.Close()
         Return
      End If

      'set title of the form.
      Dim description As String = BufferedPwModulatorCtrl1.SelectedDevice.Description
      Me.Text = "Buffered PWM Output - Run(" & description & ")"

      btnStart.Enabled = True
      btnStop.Enabled = False

   End Sub

   Private Shared Sub ShowErrorMessage(ByVal e As Exception)
      Dim errorInfo As String
      errorInfo = "There's some error happened, the error information: "
      MessageBox.Show(errorInfo & e.Message)
   End Sub

   Private Sub HandleError(ByVal err As ErrorCode)
      If err <> ErrorCode.Success Then
         MessageBox.Show("Sorry ! There are some errors happened, the error code is: " & err.ToString(), "BuffererPWMOutput")
      End If
   End Sub

   Private Sub Start_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
      Dim err As ErrorCode = ErrorCode.Success
      m_channelCountMax = BufferedPwModulatorCtrl1.Features.ChannelCountMax
      m_channelEnable = New Byte(m_channelCountMax) {}
      m_channelEnable = BufferedPwModulatorCtrl1.ChannelMap

      m_sampleCount = BufferedPwModulatorCtrl1.Samples
      m_intervelCount = BufferedPwModulatorCtrl1.IntervalCount

      Dim dataBuf(m_sampleCount - 1) As Automation.BDaq.PulseWidth
      For i As Integer = 0 To m_sampleCount - 1
         If (i And 1) = 0 Then
            dataBuf(i).HiPeriod = Val(data1_HiPeriod.Text)
            dataBuf(i).LoPeriod = Val(data1_LoPeriod.Text)
         Else
            dataBuf(i).HiPeriod = Val(data2_HiPeriod.Text)
            dataBuf(i).LoPeriod = Val(data2_LoPeriod.Text)
         End If
      Next

      BufferedPwModulatorCtrl1.IntervalCount = m_intervelCount

      BufferedPwModulatorCtrl1.Samples = m_sampleCount

      BufferedPwModulatorCtrl1.Streaming = True

      For j As Integer = 0 To m_channelCountMax - 1
         If m_channelEnable(j) = 1 Then
            BufferedPwModulatorCtrl1.SetData(j, m_sampleCount, dataBuf)
         End If
      Next

      err = BufferedPwModulatorCtrl1.Prepare()
      If err <> ErrorCode.Success Then
         HandleError(err)
         Return
      End If
      ' Enable execution status
      ExecutionStatus.Enabled = True
      Try
         BufferedPwModulatorCtrl1.Enabled = True
      Catch ex As Exception
         ShowErrorMessage(ex)
      End Try

      btnStop.Enabled = True
      btnStart.Enabled = False
   End Sub

   Private Sub Stop_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
      Try
         BufferedPwModulatorCtrl1.Enabled = False
      Catch ex As Exception
         ShowErrorMessage(ex)
      End Try
      ' Disable execution status
      ExecutionStatus.Enabled = False
      btnStart.Enabled = True
      btnStop.Enabled = False
   End Sub
End Class


